src/Core/KeyManagement/UserKey/Implementations/RotateUserAccountkeysCommand.cs

```diff
@@ -79,25 +86,110 @@ public async Task<IdentityResult> RotateUserAccountKeysAsync(User user, RotateUs
         user.LastKeyRotationDate = now;
         user.SecurityStamp = Guid.NewGuid().ToString();
 
-        if (
-            !model.MasterPasswordUnlockData.ValidateForUser(user)
-        )
+        List<UpdateEncryptedDataForKeyRotation> saveEncryptedDataActions = [];
+
+        await UpdateAccountKeys(model, user, saveEncryptedDataActions);
+        UpdateUnlockMethods(model, user, saveEncryptedDataActions);
+        UpdateUserData(model, user, saveEncryptedDataActions);
+
+        await _userRepository.UpdateUserKeyAndEncryptedDataV2Async(user, saveEncryptedDataActions);
+        await _pushService.PushLogOutAsync(user.Id);
+        return IdentityResult.Success;
+    }
+
+    async Task<bool> IsV2EncryptionUserAsync(User user)
+    {
+        // A V2 user has a signature key pair, and their user key is COSE
+        // The user key cannot be directly checked here; but the items encrypted with it can be checked.
+        ArgumentNullException.ThrowIfNull(user);
+        var hasSignatureKeyPair = await _userSignatureKeyPairRepository.GetByUserIdAsync(user.Id) != null;
+        var isPrivateKeyEncryptionV2 = GetEncryptionType(user.PrivateKey) == EncryptionType.XChaCha20Poly1305_B64;
+
+        // Valid v2 user
+        if (hasSignatureKeyPair && isPrivateKeyEncryptionV2)
         {
-            throw new InvalidOperationException("The provided master password unlock data is not valid for this user.");
+            return true;
         }
-        if (
-            model.AccountPublicKey != user.PublicKey
-        )
+
+        // Valid v1 user
+        if (!hasSignatureKeyPair && !isPrivateKeyEncryptionV2)
+        {
+            return false;
+        }
+
+        throw new InvalidOperationException("User is in an invalid state for key rotation. User has a signature key pair, but the private key is not in v2 format, or vice versa.");
+    }
+
+    public async Task ValidateRotationModelSignatureKeyPairForV2User(RotateUserAccountKeysData model, User user)
+    {
+        var currentSignatureKeyPair = await _userSignatureKeyPairRepository.GetByUserIdAsync(user.Id);
+        if (model.AccountKeys == null || model.AccountKeys.SignatureKeyPairData == null)
+        {
+            throw new InvalidOperationException("The provided signing key data is null, but the user already has signing keys.");
+        }
+        if (model.AccountKeys.SignatureKeyPairData.VerifyingKey != currentSignatureKeyPair.VerifyingKey)
+        {
+            throw new InvalidOperationException("The provided verifying key does not match the user's current verifying key.");
+        }
+        if (string.IsNullOrEmpty(model.AccountKeys.PublicKeyEncryptionKeyPairData?.SignedPublicKey))
```

Comment:
The `PublicKeyEncryptionKeyPairData` cannot be null
```suggestion
        if (string.IsNullOrEmpty(model.AccountKeys.PublicKeyEncryptionKeyPairData.SignedPublicKey))
```